AZ CLI KEDA add-on changes (#94129)
Azure/azure-content: articles/aks/keda-deploy-add-on-cli.md

---

> title: Deploy the Kubernetes Event-driven Autoscaling (KEDA) add-on by using Azure CLI

Reviewer: ```suggestion
title: Install the Kubernetes Event-driven Autoscaling (KEDA) add-on by using Azure CLI
```

---
title: Deploy the Kubernetes Event-driven Autoscaling (KEDA) add-on by using Azure CLI
description: Use Azure CLI to deploy the Kubernetes Event-driven Autoscaling (KEDA) add-on to Azure Kubernetes Service (AKS).
author: Ramya Oruganti
ms.author: raorugan
ms.service: container-service
ms.topic: how-to article
ms.date: 06/08/2022
ms.custom: template-how-to 
---

> 
> # Install the Kubernetes Event-driven Autoscaling (KEDA) add-on by using Azure CLI

Reviewer: Just for consistency with ARM document
```suggestion
# Deploy the Kubernetes Event-driven Autoscaling (KEDA) add-on by using Azure CLI
```

> This article shows you how to install the Kubernetes Event-driven Autoscaling (KEDA) add-on to Azure Kubernetes Service (AKS) by using Azure CLI. The article includes steps to verify that it's installed and running

Reviewer: The second sentence is cut off, is this intentional?

[!INCLUDE [Current version callout](./includes/keda/current-version-callout.md)]

[!INCLUDE [preview features callout](./includes/preview/preview-callout.md)]

> 

Reviewer: Nit
```suggestion

```

## Prerequisites

- An Azure subscription. If you don't have an Azure subscription, you can create a [free account](https://azure.microsoft.com/free).

> - [Azure CLI installed](/cli/azure/install-azure-cli).

Reviewer: Any specific version of the CLI that is required?

Author: added guidance on aks extension to be upgraded if the extension is incompatible with az cli version it will throw error and user will need to upgrade the az cli appropriately

Reviewer: Wouldn't it be good to also simply list what version is required? This would be more customer-friendly

```azurecli-interactive
az provider register --namespace Microsoft.ContainerService
```

## Install the KEDA add-on with Azure CLI
To install the KEDA add-on, use `--enable-keda` when creating or updating a cluster.

The following example creates a *myResourceGroup* resource group. Then it creates a *myAKSCluster* cluster with the KEDA add-on.

```azurecli-interactive
az group create --name myResourceGroup --location eastus

az aks create \
  --resource-group myResourceGroup \
  --name myAKSCluster \
  --enable-keda 
```

For existing clusters, use `az aks update` with `--enable-keda` option. The following code shows an example.

```azurecli-interactive
az aks update \
  --resource-group myResourceGroup \
  --name myAKSCluster \
  --enable-keda 
```

## Get the credentials for your cluster

Get the credentials for your AKS cluster by using the `az aks get-credentials` command. The following example command gets the credentials for *myAKSCluster* in the *myResourceGroup* resource group:

```azurecli-interactive
az aks get-credentials --resource-group myResourceGroup --name myAKSCluster
```

## Verify that the KEDA add-on is installed on your cluster

> To see if the KEDA add-on is installed on your cluster, verify that the `enabled` value is `true` for `keda` under `workloadAutoScalerProfile`. The following example shows the status of the KEDA add-on for *myAKSCluster* in *myResourceGroup*:

Reviewer: This reads easier imo
```suggestion
To see if the KEDA add-on is installed on your cluster, verify that the `enabled` value is `true` for `keda` under `workloadAutoScalerProfile`.

The following example shows the status of the KEDA add-on for *myAKSCluster* in *myResourceGroup*:
```

```azurecli-interactive
az aks show -g "myResourceGroup" --name myAKSCluster --query "workloadAutoScalerProfile.keda.enabled" 
```
## Verify that KEDA is running on your cluster

You can verify KEDA that's running on your cluster. Use `kubectl` to display the operator and metrics server installed in the AKS cluster under kube-system namespace. For example:

```azurecli-interactive
kubectl get pods -n kube-system 
```

> The following example output shows the keda operator and keda metrics apiserver installed in the AKS cluster along with the status

Reviewer: Aligning as per https://github.com/kedacore/governance/blob/main/BRANDING.md#branding
```suggestion
The following example output shows that the KEDA operator and metrics API server are installed in the AKS cluster along with its status:
```

```output
kubectl get pods -n kube-system

keda-operator-********-k5rfv                     1/1     Running   0          43m
keda-operator-metrics-apiserver-*******-sj857   1/1     Running   0          43m
```
To verify the version of your KEDA, use `kubectl get crd/scaledobjects.keda.sh -o yaml `. For example:

```azurecli-interactive
kubectl get crd/scaledobjects.keda.sh -o yaml 
```

> The following example output shows the configuration of KEDA:

Reviewer: ```suggestion
The following example output shows the configuration of KEDA in the `app.kubernetes.io/version` label:
```

                type: integer
              resourceMetricNames:
                items:
                  type: string
                type: array
              scaleTargetGVKR:
                description: GroupVersionKindResource provides unified structure for
                  schema.GroupVersionKind and Resource
                properties:
                  group:
                    type: string
                  kind:
                    type: string
                  resource:
                    type: string
                  version:
                    type: string
                required:
                - group
                - kind
                - resource
                - version
                type: object
              scaleTargetKind:
                type: string
            type: object
        required:
        - spec
        type: object
    served: true
    storage: true
    subresources:
      status: {}
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: "2022-06-08T10:31:07Z"
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  storedVersions:
  - v1alpha1

>   ```

Reviewer: This is too much information, let's reduce it for sake of easier reading:
```suggestion
```yaml
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: "2022-06-08T10:31:06Z"
  generation: 1
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    app.kubernetes.io/component: operator
    app.kubernetes.io/name: keda-operator
    app.kubernetes.io/part-of: keda-operator
    app.kubernetes.io/version: 2.7.0
  name: scaledobjects.keda.sh
  resourceVersion: "2899"
  uid: 85b8dec7-c3da-4059-8031-5954dc888a0b
spec:
  conversion:
    strategy: None
  group: keda.sh
  names:
    kind: ScaledObject
    listKind: ScaledObjectList
    plural: scaledobjects
    shortNames:
    - so
    singular: scaledobject
  scope: Namespaced
  # Redacted for simplicity in this article
  ```
```

>   ```

Reviewer: This is too much information, let's reduce it for sake of easier reading:
```suggestion
```yaml
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: "2022-06-08T10:31:06Z"
  generation: 1
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    app.kubernetes.io/component: operator
    app.kubernetes.io/name: keda-operator
    app.kubernetes.io/part-of: keda-operator
    app.kubernetes.io/version: 2.7.0
  name: scaledobjects.keda.sh
  resourceVersion: "2899"
  uid: 85b8dec7-c3da-4059-8031-5954dc888a0b
spec:
  conversion:
    strategy: None
  group: keda.sh
  names:
    kind: ScaledObject
    listKind: ScaledObjectList
    plural: scaledobjects
    shortNames:
    - so
    singular: scaledobject
  scope: Namespaced
  # Redacted for simplicity in this article
  ```
```

>   ```

Reviewer: This is too much information, let's reduce it for sake of easier reading:
```suggestion
```yaml
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: "2022-06-08T10:31:06Z"
  generation: 1
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    app.kubernetes.io/component: operator
    app.kubernetes.io/name: keda-operator
    app.kubernetes.io/part-of: keda-operator
    app.kubernetes.io/version: 2.7.0
  name: scaledobjects.keda.sh
  resourceVersion: "2899"
  uid: 85b8dec7-c3da-4059-8031-5954dc888a0b
spec:
  conversion:
    strategy: None
  group: keda.sh
  names:
    kind: ScaledObject
    listKind: ScaledObjectList
    plural: scaledobjects
    shortNames:
    - so
    singular: scaledobject
  scope: Namespaced
  # Redacted for simplicity in this article
  ```
```

>   ```

Reviewer: This is too much information, let's reduce it for sake of easier reading:
```yaml
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.8.0
  creationTimestamp: "2022-06-08T10:31:06Z"
  generation: 1
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    app.kubernetes.io/component: operator
    app.kubernetes.io/name: keda-operator
    app.kubernetes.io/part-of: keda-operator
    app.kubernetes.io/version: 2.7.0
  name: scaledobjects.keda.sh
  resourceVersion: "2899"
  uid: 85b8dec7-c3da-4059-8031-5954dc888a0b
spec:
  conversion:
    strategy: None
  group: keda.sh
  names:
    kind: ScaledObject
    listKind: ScaledObjectList
    plural: scaledobjects
    shortNames:
    - so
    singular: scaledobject
  scope: Namespaced
  # Redacted for simplicity in this article
  ```

> KEDA provides multiple customization options. The KEDA add-on enables the basic common configuration. If you have requirement to run with some  custom configurations for example:  number of KEDA pods running, namespaces that should be watched, logging level, etc. then you may edit the KEDA yaml manually and deploy it. Remember the customizations would be managed by you. 

Reviewer: This is a bit of oddly written, what about this?

```suggestion
While KEDA provides various customization options, the KEDA add-on currently provides basic common configuration.

If you have requirement to run with additional custom configurations, such as namespaces that should be watched or tweaking the log level, then you may edit the KEDA YAML manually and deploy it.

However, when the installation is customized it will no longer be supported. 
```

Also, I've changed the last bit given this most probably means it is no longer supported? Should we add a way that they can provide feedback on what they want to configure?

## Disable KEDA add-on from your AKS cluster

When you no longer need KEDA add-on in the cluster, use the `az aks update` command with--disable-keda option. This will disable KEDA workload auto-scaler

```azurecli-interactive
az aks update \
  --resource-group myResourceGroup \
  --name myAKSCluster \
  --disable-keda 
```

> ## Troubleshooting guide

Reviewer: This section applies to the ARM-based installation as well.

We already have this defined on https://docs.microsoft.com/en-us/azure/aks/keda-about#add-on-limitations but this section is an added value, but I'd argue we need a dedication page for this similar to [OSM's troubleshooting page](https://docs.microsoft.com/en-us/azure/aks/open-service-mesh-troubleshoot)

Author: Added to ARM page as well. Since we have only one error it didnt make sense to create a page for this. As we progress we can add further potential known issues

> 

Reviewer: Let's add a title:
```suggestion

### Enabling add-on on clusters with self-managed open-source KEDA installations

```

Thoughts?

> If managed KEDA is installed in an AKS cluster with user installed KEDA, then Managed KEDA overrides user's KEDA. If user's KEDA has some customizations added, they'll get lost once managed KEDA replaces it. The installation will complete without error. User's auto-scaling should most probably work fine as before. There will be two copies of KEDA resources present in the AKS cluster and only one of them will be doing the job of scaling.

Reviewer: What about this?
```suggestion
While Kubernetes only allows one metric server to be installed, you can in theory install KEDA multiple times. However, this is not recommended given only one installation will work.

When the KEDA add-on is installed in an AKS cluster, the previous installation of open-source KEDA will be overridden and the add-on will take over.

This means that the customization and configuration of the self-installed KEDA deployment will get lost and no longer be applied.

While there is a possibility that the existing autoscaling will keep on working, there is a risk given it will be configured differently and will not support features such as managed identity.

It is recommended to uninstall existing KEDA installations before enabling the KEDA add-on given the installation will succeed without any error.
```

Following error will be thrown in the operator logs. The installation will complete without error.

Error logged in now-suppressed non-participating KEDA operator pod:

> the error logged inside the already installed KEDA operator logs.

Reviewer: I don't get what you mean here, can you clarify this please?

> E0520 11:51:24.868081 1 leaderelection.go:330] error retrieving resource lock default/operator.keda.sh: config maps "operator.keda.sh" is forbidden: User "system:serviceaccount:default:keda-operator" cannot get resource "config maps" in API group "" in the namespace "default"

Reviewer: Are you sure that this is correct?

Author: yes, 
![image](https://user-images.githubusercontent.com/45637559/172840532-9084016f-ab3f-46bd-b772-89fe256dce52.png)
1.654770958628984e+09   INFO    Starting server {"kind": "health probe", "addr": "[::]:8081"}   
E0609 10:36:09.094659       1 leaderelection.go:330] error retrieving resource lock kube-system/operator.keda.sh: rpc error: code = Unavailable desc = error reading from server: read tcp 172.31.15.19:48892->172.30.85.205:2379: read: connection timed out
I0609 10:36:34.156241       1 leaderelection.go:258] successfully acquired lease kube-system/operator.keda.sh

## Next steps
This article showed you how to install the KEDA add-on on an AKS cluster using Azure CLI. The steps to verify that KEDA add-on is installed and running are included. With the KEDA add-on installed on your cluster, you can [deploy a sample application][keda-sample] to start scaling apps

> 

Reviewer: ```suggestion

```

### Enabling add-on on clusters with self-managed open-source KEDA installations

While Kubernetes only allows one metric server to be installed, you can in theory install KEDA multiple times. However, it isn't recommended given only one installation will work.

When the KEDA add-on is installed in an AKS cluster, the previous installation of open-source KEDA will be overridden and the add-on will take over.

This means that the customization and configuration of the self-installed KEDA deployment will get lost and no longer be applied.

While there's a possibility that the existing autoscaling will keep on working, there's a risk given it will be configured differently and won't support features such as managed identity.

It's recommended to uninstall existing KEDA installations before enabling the KEDA add-on given the installation will succeed without any error.

Following error will be thrown in the operator logs but the installation of KEDA add-on will be completed. 

Error logged in now-suppressed non-participating KEDA operator pod:
the error logged inside the already installed KEDA operator logs.

> E0520 11:51:24.868081 1 leaderelection.go:330] error retrieving resource lock default/operator.keda.sh: config maps "operator.keda.sh" is forbidden: User "system:serviceaccount:default:keda-operator" can't get resource "config maps" in API group "" in the namespace "default"

Reviewer: Better to use the exact log message rather than aligning it
